### PR TITLE
Add task to structure when possible

### PR DESCRIPTION
**Pull request by vachillo**
- [x] I have read and agree to the contributing guidelines for [submitting new pull requests](https://github.com/griptape-ai/griptape?tab=readme-ov-file#submitting-pull-requests).

## Describe your changes
this adds the task to the structure that is associated with either task when using `add_parent/child` when passing a `BaseTask`

### CODE REVIEW

File: griptape/tasks/base_task.py [hunk 1]
```diff
@@ -168,3 +177,14 @@ def full_context(self) -> dict[str, Any]:
             return structure_context
         else:
             return {}
+
+    def _attach_to_structure(self, task: BaseTask):
```
Comment (collindutter) [inline griptape/tasks/base_task.py:181]: What do you think about renaming to `_add_to_structure`? "attach" conveys more of a 1-to-1 relationship.

File: griptape/utils/__init__.py [hunk 2]
```diff
@@ -23,6 +23,10 @@ def minify_json(value: str) -> str:
     return json.dumps(json.loads(value), separators=(",", ":"))
 
 
+def not_none(value):
+    return value is not None
+
+
```
Comment (collindutter) [inline griptape/utils/__init__.py:29]: I don't think we should create a utility method for this.

**Review by collindutter — CHANGES_REQUESTED** [review]

File: griptape/utils/__init__.py (hunk 2 above)
Comment (vachillo) [reply griptape/utils/__init__.py:29]: was hoping to sneak it past you

File: griptape/tasks/base_task.py [hunk 3]
```diff
@@ -76,20 +76,26 @@ def add_parents(self, parents: list[str | BaseTask]) -> None:
 
     def add_parent(self, parent: str | BaseTask) -> None:
         parent_id = parent if isinstance(parent, str) else parent.id
+        parent_task = parent if isinstance(parent, BaseTask) else None
 
         if parent_id not in self.parent_ids:
             self.parent_ids.append(parent_id)
 
+        self._add_to_structure(parent_task)
```
Comment (collindutter) [inline griptape/tasks/base_task.py:84]: Won't this add a `None` Task if a parent id is provided?

File: griptape/tasks/base_task.py [hunk 4]
```diff
@@ -168,3 +174,15 @@ def full_context(self) -> dict[str, Any]:
             return structure_context
         else:
             return {}
+
+    def _add_to_structure(self, task: Optional[BaseTask]):
```
Comment (collindutter) [inline griptape/tasks/base_task.py:178]: Missing return type

File: griptape/tasks/base_task.py [hunk 5]
```diff
@@ -168,3 +174,15 @@ def full_context(self) -> dict[str, Any]:
...
+        if task is not None:
+            if self.structure is not None and task.structure is not None and self.structure != task.structure:
+                raise ValueError("Task is already associated with a different structure.")
+            elif self.structure:
```
Comment (collindutter) [inline griptape/tasks/base_task.py:182]: Use `is not None` check

File: griptape/tasks/base_task.py [hunk 6]
```diff
@@ -168,3 +174,15 @@ def full_context(self) -> dict[str, Any]:
...
+                if task not in self.structure.tasks:
+                    self.structure.add_task(task)  # pyright: ignore reportArgumentType
```
Comment (collindutter) [inline griptape/tasks/base_task.py:184]: Why the pyright ignore?

File: griptape/tasks/base_task.py [hunk 7]
```diff
@@ -168,3 +174,15 @@ def full_context(self) -> dict[str, Any]:
...
+            elif task.structure:
```
Comment (collindutter) [inline griptape/tasks/base_task.py:185]: Use explicit `is not None` check

File: griptape/tasks/base_task.py [hunk 8]
```diff
@@ -168,3 +174,15 @@ def full_context(self) -> dict[str, Any]:
...
+                if self not in task.structure.tasks:
+                    self.preprocess(task.structure)
```
Comment (collindutter) [inline griptape/tasks/base_task.py:187]: Doesn't `Structure.add_task` call `preprocess`?

**Review by collindutter — CHANGES_REQUESTED** [review]

File: griptape/tasks/base_task.py (hunk 6 above)
Comment (vachillo) [reply griptape/tasks/base_task.py:184]: it says that `BaseTask | None` is not valid for `add_task`. even though theres a None check

File: griptape/tasks/base_task.py (hunk 8 above)
Comment (vachillo) [reply griptape/tasks/base_task.py:187]: there was a reason this was here...seems like its not needed now.

File: griptape/tasks/base_task.py (hunk 5 above)
Comment (vachillo) [reply griptape/tasks/base_task.py:182]: is there a linting rule that can be added for this?

File: griptape/tasks/base_task.py (hunk 6 above)
Comment (vachillo) [reply griptape/tasks/base_task.py]: oh it passes now

File: tests/unit/structures/test_workflow.py
Comment (collindutter) [inline tests/unit/structures/test_workflow.py:1]: Can we leave this test as is?

Comment (collindutter) [inline tests/unit/structures/test_workflow.py:1]: I think better to leave the existing tests as is and add new tests for each of the Structures.

Comment (collindutter) [inline tests/unit/structures/test_workflow.py:1]: Can you also add tests for Pipeline and Agent?

**Pull request closed without merging**